cpp/src/arrow/dataset/file_base.cc

```diff
@@ -160,6 +160,13 @@ Status FileWriter::Write(RecordBatchReader* batches) {
   return Status::OK();
 }
 
+Status FileWriter::Finish() {
+  RETURN_NOT_OK(FinishInternal());
+  return destination_->Close();
```

Comment:
Note: explicitly closing output files is preferrable, especially with remote filesystems where this might plausibly fail.

Reply:
(and MockFileSystem will deliberately not write anything out if you don't close it explicitly)